sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OnHeapCachedBatch.java

```diff
@@ -0,0 +1,403 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.vectorized;
+
+import java.nio.ByteBuffer;
+
+import org.apache.spark.memory.MemoryMode;
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow;
+import org.apache.spark.sql.catalyst.expressions.codegen.BufferHolder;
+import org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter;
+import org.apache.spark.sql.execution.columnar.*;
+import org.apache.spark.sql.types.*;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * A column backed by an in memory JVM array.
+ */
+public final class OnHeapCachedBatch extends ColumnVector implements java.io.Serializable {
+
+  // keep compressed data
+  private byte[] buffer;
+
+  // whether a row is already extracted or not. If extractTo() is called, set true
+  // e.g. when isNullAt() and getInt() ara called, extractTo() must be called only once
+  private boolean[] calledExtractTo;
+
+  // a row where the compressed data is extracted
+  private transient UnsafeRow unsafeRow;
+  private transient BufferHolder bufferHolder;
+  private transient UnsafeRowWriter rowWriter;
+  private transient MutableUnsafeRow mutableRow;
+
+  // accesssor for a column
+  private transient ColumnAccessor columnAccessor;
+
+  // an accessor uses only column 0
+  private final int ORDINAL = 0;
+
+  protected OnHeapCachedBatch(int capacity, DataType type) {
+    super(capacity, type, MemoryMode.ON_HEAP_CACHEDBATCH);
+    reserveInternal(capacity);
+    reset();
+  }
+
+  @Override
+  public long valuesNativeAddress() {
+    throw new RuntimeException("Cannot get native address for on heap column");
+  }
+  @Override
+  public long nullsNativeAddress() {
+    throw new RuntimeException("Cannot get native address for on heap column");
+  }
+
+  @Override
+  public void close() {
+  }
+
+  private void initialize() {
+    if (columnAccessor == null) {
+      setColumnAccessor();
+    }
+    if (mutableRow == null) {
+      setRowSetter();
+    }
+  }
+
+  private void setColumnAccessor() {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(buffer);
+    columnAccessor = ColumnAccessor$.MODULE$.apply(type, byteBuffer);
+    calledExtractTo = new boolean[capacity];
+  }
+
+  private void setRowSetter() {
+    unsafeRow = new UnsafeRow(1);
+    bufferHolder = new BufferHolder(unsafeRow);
+    rowWriter = new UnsafeRowWriter(bufferHolder, 1);
+    mutableRow = new MutableUnsafeRow(rowWriter);
+  }
+
+  // call extractTo() before getting actual data
+  private void prepareRowAccess(int rowId) {
```

Comment:
This PR is building the infrastructure that not being used yet, so I think we don't need to rush.